Fix rotation calculations. Add tests.

TimAbraldes opened the pull request:
The j2me API has far fewer footguns than the HTML Canvas API.
Implementing it using the HTML Canvas API involves a lot of severed
feet.

WIP because tests are failing

TimAbraldes commented:
Most tests are now fixed. The remaining issue is that, for some reason, strings are drawn with a bigger font

TimAbraldes commented:
Turns out the font thing was only happening on my machine. Travis was failing due to pixel differences but manual inspection showed that the images were identical-enough-that-I-couldn't-tell-a-difference. Added thresholds based on the pixel differences that Travis reported

TimAbraldes commented:
No longer WIP

marco-c commented:
Could you use the red rectangle image instead of the Firefox logo? This should make the number of different pixels much smaller (I think 3000 different pixels are so many that we won't be able to catch regressions).

TimAbraldes commented:
> Could you use the red rectangle image instead of the Firefox logo? This should make the number of different pixels much smaller (I think 3000 different pixels are so many that we won't be able to catch regressions).

A red rectangle by itself isn't super useful for testing (e.g.) mirrored rotation. I can use a test image that I create on the fly; it will be a rectangle that is larger in one dimension than the other (probably taller than it is wide) with distinctly-colored rectangles in each of its 4 corners

TimAbraldes commented:
Updated tests to use solid-color images instead of the firefox logo. Let's see how the Travis run turns out!

TimAbraldes commented:
Hmmm.... TransMirrorRot90AnchorVCenterRight and TransMirrorRot90AnchorBottomRight don't look right

marco-c commented:
For DrawRegionTransMirrorRot180AnchorVCenterHCenter, this is the diff between the reference image and the image in the Travis build:
![diff](https://cloud.githubusercontent.com/assets/1616846/6952970/79123a32-d8c6-11e4-9d21-7357804cc998.png)

Reference image:
![image1](https://cloud.githubusercontent.com/assets/1616846/6952973/83c2569c-d8c6-11e4-89c1-5540152d0fd2.png)


Image from the Travis build:
![image2](https://cloud.githubusercontent.com/assets/1616846/6952975/86b79c86-d8c6-11e4-8b73-99a125e32b09.png)

TimAbraldes commented:
Rebased and fixed some lingering issues. Also added tests for the remaining rotation/anchor combinations. Note that we've passed Travis' log limit of 10,000 lines. Not sure what to do there.

marco-c commented on midp/gfx.js:
```diff
@@ -1178,38 +1135,114 @@ var currentlyFocusedTextEditor;
+        // Make `absX` and `absY` the top-left coordinates where we will
+        // place the image in absolute coordinates
+        if (0 != (anchor & HCENTER)) {
+            absX -= (w / 2);
```
Can you use *(w >>> 1) | 0*? Maybe it's the absence of truncation that is generating the small differences in the rendering?

TimAbraldes commented:
Added a changeset that switches to ((w >>> 1) | 0) and ((h >>> 1) | 0)

marco-c commented:
I guess the images need to be re-generated now.

TimAbraldes commented:
Startup benchmark 100 rounds, 10s round delay:

|      Test     | Baseline Mean |   Mean   |  +/-  |   %   |       P       |    Min   |    Max   | 
|:--------------|--------------:|---------:|------:|------:|--------------:|---------:|---------:|
| startupTime   |       1,641ms |  1,624ms | -17ms | -1.02 | INSIGNIFICANT |  1,537ms |  1,831ms | 
| totalSize     |      28,894kb | 29,022kb | 128kb |  0.44 | INSIGNIFICANT | 27,479kb | 30,968kb | 
| domSize       |         122kb |    122kb |   0kb | -0.00 | INSIGNIFICANT |    122kb |    123kb | 
| styleSize     |         435kb |    435kb |   0kb |  0.08 | INSIGNIFICANT |    435kb |    448kb | 
| jsObjectsSize |      17,371kb | 17,388kb |  17kb |  0.10 |         WORSE | 17,372kb | 17,434kb | 
| jsStringsSize |         666kb |    720kb |  54kb |  8.18 |         WORSE |    719kb |    725kb | 
| jsOtherSize   |       9,936kb |  9,990kb |  54kb |  0.54 | INSIGNIFICANT |  8,460kb | 11,940kb | 
| otherSize     |         363kb |    365kb |   2kb |  0.61 | INSIGNIFICANT |    356kb |    447kb |
